include/glow/Backends/DeviceManager.h

```diff
@@ -0,0 +1,111 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_BACKENDS_DEVICEMANAGER_H
+#define GLOW_BACKENDS_DEVICEMANAGER_H
+
+#include "glow/Backends/Backend.h"
+#include "glow/Backends/CompiledFunction.h"
+#include "glow/Graph/Context.h"
+#include "glow/Graph/Graph.h"
+#include "glow/Support/ThreadPool.h"
+
+#include <atomic>
+#include <condition_variable>
+#include <map>
+#include <mutex>
+#include <thread>
+#include <unordered_map>
+
+namespace glow {
+
+using DeviceNetworkID = size_t;
+
+enum ResultCode { READY, EXECUTED, FAILED, CANCELLED };
+
+using ReadyCB = std::function<void(const Module *, ResultCode)>;
+using RunIdentifier = uint64_t;
+using ResultCB =
+    std::function<void(RunIdentifier, ResultCode, std::unique_ptr<Context>)>;
+using FunctionMap = std::map<std::string, CompiledFunction *>;
+
+class DeviceManager {
+protected:
+  // Type of Backend for this Device.
+  BackendKind backend_;
+
+  /// Thread which interfaces with the device.
+  ThreadPool workThread_;
```

Comment:
It would be nice if we could hide the implementation details like `workThread` and `nextIdentifier`, so that only the implementation C++ file sees them. Then we could remove the includes for thread pools and atomics from this header file.

But C++ does not make it easy ;-)

Reply:
We could do something similar to the backend where DeviceManager is an abstract interface and you must create them via a factory function. Hides the implementation details, but adds a new level of indirection. I'm guessing since thats what we do for backends thats fine though.

Reply:
Actually I'm not sure we can and have those parts be inheritable by subclasses. I can move those members to the CPUDeviceManager which would clean up the DeviceManager header, but then we'd need to copy them if another device wants the ThreadPool, etc. What do you think?

Reply:
My main concern why I wanted these details to be hidden is: Are we sure that all device managers would always use <thread>-based implementation for their threading? Is it e.g. possible that in some cases a thread pool is provided by Caffe2 or any other outer system that uses Glow as a backend?

The second, less important concern is: Only the API is important. The clients of DeviceManager do not care much about how it functions internally and should not see those details.

> I can move those members to the CPUDeviceManager which would clean up the DeviceManager header, but then we'd need to copy them if another device wants the ThreadPool, etc. 

If you go along this lines, you could define a helper class with common data and logic that can be used by all interested implementations of device managers. And then those concrete DeviceManager implementations could include it as a member or inherit from it.

Reply:
Ah right, yeah thats easy enough. Will do.